src/main/java/org/sagebionetworks/bridge/services/PermissionService.java

```diff
@@ -158,6 +172,120 @@ public void deletePermission(String appId, String guid) {
         permissionDao.deletePermission(appId, guid);
     }
     
+    public void updatePermissionsFromRoles(Account account, Account persistedAccount) {
+        checkNotNull(account);
+        checkNotNull(persistedAccount);
+        
+        Map<EntityType, Map<String, Set<AccessLevel>>> newPermissions = getPermissionsForRoles(account);
+        Map<EntityType, Map<String, Set<AccessLevel>>> oldPermissions = getPermissionsForRoles(persistedAccount);
+        
+        Map<EntityType, Map<String, Set<AccessLevel>>> additionalPermissions = new HashMap<>();
+        Map<EntityType, Map<String, Set<AccessLevel>>> removablePermissions = new HashMap<>();
+        
+        for (EntityType entityType : newPermissions.keySet()) {
+            if (oldPermissions.containsKey(entityType)) {
+                // check which permissions are already accounted for, add any missing
+                Map<String, Set<AccessLevel>> newPermissionEntityTypeMap = newPermissions.get(entityType);
+                for (String entityId : newPermissionEntityTypeMap.keySet()) {
+                    if (oldPermissions.get(entityType).containsKey(entityId)) {
+                        Set<AccessLevel> oldAccessLevels = oldPermissions.get(entityType).get(entityId);
+                        Set<AccessLevel> additionalAccessLevels = new HashSet<>();
+                        for (AccessLevel accessLevel : newPermissionEntityTypeMap.get(entityId)) {
+                            if (!oldAccessLevels.contains(accessLevel)) {
+                                additionalAccessLevels.add(accessLevel);
+                            }
+                        }
+                        Map<String, Set<AccessLevel>> entityTypeMap = additionalPermissions.getOrDefault(entityType, new HashMap<>());
+                        entityTypeMap.put(entityId, additionalAccessLevels);
+                        additionalPermissions.put(entityType, entityTypeMap);
+                    } else {
+                        Map<String, Set<AccessLevel>> entityTypeMap = additionalPermissions.getOrDefault(entityType, new HashMap<>());
+                        entityTypeMap.put(entityId, newPermissionEntityTypeMap.get(entityId));
+                        additionalPermissions.put(entityType, entityTypeMap);
+                    }
```

Comment:
I didn't run coverage! I'm doing that now. And you're right that this can't be hit. I'm counting on the orgId staying the same in other places. So I'll remove this block.